src/main/java/com/boot/cart/service/CartService.java

```diff
@@ -1,255 +1,251 @@
-            throw new InvalidInputDataException(
-                    "You currently have " + productsInCart + " Products: " + productName + " in cart!");
-        }
+        List<CartEntry> cartEntries;
+        double productTotal = 0;
 
-        if (Math.toIntExact(productsInCart) < quantity) {
-            throw new InvalidInputDataException("You cannot remove more than " + productsInCart + " " + productName
-                    + " Products from the shopping cart!");
-        } else {
-            double productTotal = 0;
+            cartEntries = cart.getEntries();
+            CartEntry cartEntry = cartEntries.stream().filter(entry -> productName.equals(entry.getProductName())).findFirst().orElse(null);
 
-            for (int i = 0; i < quantity; i++) {
-                productList.remove(product);
-                productTotal += product.getProductPrice();
+            if (cartEntry == null) {
+                throw new EntityNotFoundException("CartEntry not found in the Database!");
+            }
+            else{
+                cartEntry.setQuantity(cartEntry.getQuantity() - quantity);
+                cartEntry.setCart(cart);
             }
 
-            cart.setUser(user);
+        for (int i = 0; i < quantity; i++) {
+            productTotal -= productDTO.getPrice();
+        }
 
-            cart.setProductList(productList);
-            cart.setLastUpdatedOn(LocalDateTime.now());
-            cart.setTotal(cart.getTotal() - productTotal);
+        cart.setEntries(cartEntries);
+        cart.setTotal(cart.getTotal() + productTotal);
 
-            cartRepository.save(cart);
+        cartRepository.save(cart);
 
-            return CartMapper.cartEntityToDto(cart);
-        }
+        return cartEntityToDto(cartRepository.findByUserId(user.getId()));
     }
 
     public void deleteCartByEmail(String email) throws  EntityNotFoundException {
-
-        User user;
+        UserDTO user;
         try {
-            user = UserMapper.DtoToUserEntity(userServiceClient.callGetUserByEmail(email));
+            user = userServiceClient.callGetUserByEmail(email);
         } catch (HttpClientErrorException.NotFound e) {
             throw new EntityNotFoundException("Email: " + email + " not found in the Database!");
         }
-
-        //TODO you might benefit from hibernate caching but you might also do the same DB query twice here(once when you want to see it's not null and 2nd time when you use it)
-        if (cartRepository.findByUser(user) != null) {
-            Cart cart = cartRepository.findByUser(user);
-
-            List<Product> prodList = cart.getProductList();
-
-            Iterator<Product> iter = prodList.iterator();
-
-            while (iter.hasNext()) {
-                Product product = iter.next();
-                iter.remove();
-
-                log.info("{} succesfully deleted from Product List",product.getProductName());
-
-            }
+        Cart cart = cartRepository.findByUserId(user.getId());
```

Comment:
you could validate that the cart was found

Reply:
Done!